### PR TITLE
Merge security logic 

### PR DESCRIPTION
**Pull request by larat7**
Merge the `security-merge` branch into `master` to avoid having to keep track of both of them.
Currently it passes the same benchmarks and tests as `master`.

### CODE REVIEW

**Comment by jonhoo** [comment]
It looks like at least one test fails on Travis: https://travis-ci.org/mit-pdos/distributary/builds/347553194#L949

**Comment by ms705** [comment]
Travis runs `cargo test --all`, and hence caught this. 

From a quick look, the change originates in c2ed208ca, which adds code for creating group context nodes. I suspect the generation of the group context node name converts a `DataType` to string, but isn't supposed to quote it because that string get concatenated with others to form the name.

**Comment by larat7** [comment]
Yeah, I often need to format node names for specific universes and treating `DataType` as an unquoted string makes it way more convenient. Any suggestions?

File: dataflow/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ tarpc = "0.9.0"
 tarpc-plugins = "0.2.0"
 tokio-core = "0.1.8"
 vec_map = { version = "0.8.0", features = ["eders"] }
+hyper = "0.11.6"
```
Comment (ms705) [inline dataflow/Cargo.toml]: Why is this necessary? EDIT: ah, for trigger nodes' HTTP requests.

File: mir/src/node.rs [hunk 2]
```diff
@@ -227,10 +227,9 @@ impl MirNode {
             // otherwise, just look up in the column set
             _ => match self.columns
                 .iter()
-                .position(|cc| cc.name == c.name && cc.table == c.table)
+                .position(|cc| cc.name == c.name)
```
Comment (ms705) [inline mir/src/node.rs]: Is this relaxation potentially dangerous? Seems like we could end up with the wrong column if two tables we join had the a column of the same name. (note to self: check with `git blame` why we added this.)

File: mir/src/optimize.rs [hunk 3]
```diff
@@ -10,10 +10,11 @@ pub fn optimize(q: MirQuery) -> MirQuery {
     q
 }
 
-pub fn optimize_post_reuse(q: &mut MirQuery) {
-    find_and_merge_filter_chains(q);
+pub fn optimize_post_reuse(_q: &mut MirQuery) {
+    // find_and_merge_filter_chains(q);
```
Comment (ms705) [inline mir/src/optimize.rs]: Why did we disable this optimization? I guess it affects reuse in security policies?

File: mir/src/rewrite.rs [hunk 4]
```diff
@@ -32,7 +32,7 @@ pub fn pull_required_base_columns(q: &mut MirQuery) {
                 !mn.borrow()
                     .ancestors()
                     .iter()
-                    .any(|a| a.borrow().columns().contains(c))
+                    .any(|a| a.borrow().columns().iter().any(|ac|  ac.name == c.name))
```
Comment (ms705) [inline mir/src/rewrite.rs]: This also relaxes to comparison by column name only, rather than by column name and table name. Check why we changed this.

File: src/controller/inner.rs [hunk 5]
```diff
@@ -469,6 +508,11 @@ impl ControllerInner {
         })
     }
 
+    fn get_getter(&self, name: &str) -> Option<RemoteGetter> {
```
Comment (ms705) [inline src/controller/inner.rs]: What's the purpose of this? I guess it would get a getter and immediately build it on the server, rather than expecting the owner of the `ControllerHandle` to build it?

File: src/controller/mir_to_flow.rs [hunk 6]
```diff
@@ -560,15 +599,15 @@ pub(crate) fn make_join_node(
     // at this point in the codebase, so the `r2.a = r1.b` will join on the wrong `a` column.
     let left_join_col_id = projected_cols_left
         .iter()
-        .position(|lc| lc == on_left.first().unwrap())
+        .position(|lc| lc.name == on_left.first().unwrap().name)
```
Comment (ms705) [inline src/controller/mir_to_flow.rs]: Column name/table name relaxation, see above.

File: src/controller/mir_to_flow.rs [hunk 7]
```diff
@@ -560,15 +599,15 @@ pub(crate) fn make_join_node(
...
         .expect(&format!(
             "missing left-side join column {:?} in {:?}",
             on_left.first().unwrap(),
             projected_cols_left
         ));
     let right_join_col_id = projected_cols_right
         .iter()
-        .position(|rc| rc == on_right.first().unwrap())
+        .position(|rc| rc.name == on_right.first().unwrap().name)
```
Comment (ms705) [inline src/controller/mir_to_flow.rs]: Column name/table name relaxation, see above.

**Comment by ms705** [comment]
This looks great! I left some comments for us to chase, but from what I can tell, there are no fundamental obstacles to merging. The main correctness question lingers around the changed semantics for matching columns in many places (by column name only, not by table name now).

This PR actually also splits the behemoth file in `src/controller/sql/mir.rs` into more manageable submodules, a nice side effect.

File: core/src/data.rs [hunk 8]
```diff
@@ -55,6 +55,27 @@ impl DataType {
             DataType::Timestamp(ts) => json!(ts.format("%+").to_string()),
         }
     }
+    
+    pub fn to_string(&self) -> String {
```
Comment (jonhoo) [inline core/src/data.rs]: I think this method should instead be an `impl Display for DataType`. That'd be useful in other places too!

File: core/src/data.rs [hunk 9]
```diff
@@ -223,6 +244,16 @@ impl Into<i64> for DataType {
...
 
+impl Into<i32> for DataType {
+    fn into(self) -> i32 {
+        if let DataType::Int(s) = self {
+            s
+        } else {
+            unreachable!();
+        }
+    }
+}
```
Comment (jonhoo) [inline core/src/data.rs]: Separately from this PR, I wish we had a nicer way to deal with all these `DataType` conversions. We may want a macro that generates `TryFrom` and `TryInto` conversions.

File: dataflow/src/backlog/mod.rs [hunk 10]
```diff
@@ -181,6 +181,12 @@ impl SingleReadHandle {
     pub fn len(&self) -> usize {
         self.handle.len()
     }
+
+    pub fn nrows(&self) -> usize {
```
Comment (jonhoo) [inline dataflow/src/backlog/mod.rs]: This is potentially a *very* costly operation, in part because it will also hold up writers until all the rows have been iterated over (or, rather, all the keys). May want to add a comment to that effect. Or maybe even call it `count_rows` to make it explicit that it does compute.

File: dataflow/src/node/debug.rs [hunk 11]
```diff
@@ -107,6 +112,14 @@ impl Node {
         s
     }
 
+    fn is_security(name: &str) -> bool {
+        use regex::Regex;
+
+        let re = Regex::new(r"^sp_").unwrap();
+
+        re.is_match(name)
```
Comment (jonhoo) [inline dataflow/src/node/debug.rs]: Why not `name.starts_with("sp_")` ?

File: dataflow/src/ops/rewrite.rs [hunk 12]
```diff
@@ -0,0 +1,240 @@
+use std::collections::{HashMap, HashSet};
+use prelude::*;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Rewrite {
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I'd like to see some basic `///` documentation for this operator. A basic outline of what its forward processing does using the variable names in `new()` would go a long way.

File: dataflow/src/ops/rewrite.rs [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+            src: src.into(),
+            should_rewrite: should_rewrite.into(),
+            rw_col: rw_col,
+            value: value,
+            key: key,
+        }
+    }
+
+    fn rewrite(&self, r: Vec<DataType>) -> Vec<DataType> {
+        let mut r = r.clone();
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: This seems unnecessarily expensive given that we own `r`. Should be:
```rust
fn rewrite(&self, mut r: Vec<DataType>) -> Vec<DataType> {
    r[self.rw_col] = self.value.clone();
    r
}
```

File: dataflow/src/ops/rewrite.rs [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+        r[self.rw_col] = self.value.clone();
+        r
+    }
+
+    fn generate_row(&self, r: Vec<DataType>, positive: bool) -> Record {
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I believe `(Vec<DataType>, bool)` already implements `Into<Record>`, so wherever you call this method, you should be able to just use `.into()` instead :)

File: dataflow/src/ops/rewrite.rs [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        if positive {
+            Record::Positive(r.clone())
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: The `clone()` shouldn't be necessary here since we already own `r`.

File: dataflow/src/ops/rewrite.rs [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+        vec![self.src.as_global(), self.should_rewrite.as_global()]
+    }
+
+    fn is_join(&self) -> bool {
+        true
+    }
+
+
+    fn must_replay_among(&self) -> Option<HashSet<NodeIndex>> {
+        Some(Some(self.src.as_global()).into_iter().collect())
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: Just for future reference, you can use `std::iter::once(self.src.as_global())` if you want an iterator that yields just one value :) Not important to change here though.

File: dataflow/src/ops/rewrite.rs [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        _: Option<usize>,
+        nodes: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        assert!(from == *self.src || from == *self.should_rewrite);
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I'd make this a `debug_assert`, but probably won't really affect performance.

File: dataflow/src/ops/rewrite.rs [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+                    *self.should_rewrite,
+                    &[0],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+
+                if rc.is_none() {
+                    // todo (larat): might need to do some stuff here
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: This is if you miss in the `should_rewrite` state. You should add that key to `misses`. If you don't that key will never be replayed, and those records will just never be emitted.

File: dataflow/src/ops/rewrite.rs [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+        let mut emit_rs = Vec::with_capacity(rs.len());
+
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        for r in rs {
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: It would be pretty nice if we could *just* use `rs` here and not `emit_rs`, but it's probably not worth spending time on for this PR.

File: dataflow/src/ops/rewrite.rs [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+                let other_rows = self.lookup(
+                    *self.src,
+                    &[self.key],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+                if other_rows.is_none() {
+                    // todo (larat): might need to do some stuff here
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: Update `misses`.

File: dataflow/src/ops/rewrite.rs [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+                    continue;
+                }
+
+                let mut other_rows = other_rows.unwrap().peekable();
+                if other_rows.peek().is_none() {
+                    continue;
+                } else {
+                    let mut other;;
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I would just move the `let outer` into the loop.

File: dataflow/src/ops/rewrite.rs [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+                    while other_rows.peek().is_some() {
+                        other = other_rows.next().unwrap();
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: Given that you always use `other.to_vec` below, I'd just do:
```rust
let other = other_rows.next().unwrap().to_vec();
```
And then `clone()` the thing you pass to the first call to `generate_row` below (or `.into()` if you switch to using that).

File: dataflow/src/ops/rewrite.rs [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+                        if r.is_positive() {
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I think you could get rid of the branch here and just use `!r.is_positive` as the argument to the first `generate_row`, and `r.is_positive()` as the argument to the second.

File: dataflow/src/ops/trigger.rs [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+use prelude::*;
+
+use hyper::Client;
+use tokio_core::reactor::Core;
+use serde::Serialize;
+use serde_json;
+use std::thread;
+
+/// Applies the identity operation to the view. Since the identity does nothing,
+/// it is the simplest possible operation. Primary intended as a reference
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: I think this comment is a little misleading. The operator does not *just* apply the identify operator, and arguably that's not even it's primary purpose. It'd be better if we here gave a clear description of what this operators is used for.

File: dataflow/src/ops/trigger.rs [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+    us: Option<IndexPair>,
+    src: IndexPair,
+    trigger: TriggerType,
+    key: Vec<usize>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum TriggerType {
+    // url to the blender, group name
+    GroupCreation{ url: String, group: String },
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: I would call the `url` field `controller_url`, and then get rid of the comment.

File: dataflow/src/ops/trigger.rs [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+impl Trigger {
+    /// Construct a new Trigger operator.
+    pub fn new(src: NodeIndex, trigger: TriggerType, key: Vec<usize>) -> Trigger {
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: What is the `key` parameter for? I guess this might become clearer when the docs for the `Trigger` operator are expanded a little.

File: dataflow/src/ops/trigger.rs [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+        assert_eq!(key.len(), 1);
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: If it has to be length 1, why not take `key: usize`?

File: core/src/data.rs (hunk 8 above)
Comment (ms705) [reply core/src/data.rs]: It deliberately isn't, actually -- we have such an `impl Display` already, but it quotes string literals. If we wish to use them as part of other strings (e.g., specifically, node names for group universes), we don't want quoting. Hence this additional method. (This was why a test failed earlier.)

File: dataflow/src/ops/trigger.rs [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+    ) where Q: Send {
+        use hyper;
+        let url = format!("{}/{}", url, path);
+        let requests: Vec<hyper::Request> = requests.iter().map(|req| {
+            let mut r = hyper::Request::new(hyper::Method::Post, url.clone().parse().unwrap());
+            r.set_body(serde_json::to_string(&req).unwrap());
+            r
+        }).collect();
+
+        thread::spawn(move || {
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: This is a little unfortunate, but I don't have a better way of doing it. Maybe we could instead have a long-running thread with a channel that we send requests to instead? May not be worth dealing with in this PR, but perhaps leave a `TODO` comment?

File: dataflow/src/ops/trigger.rs [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+            let mut core = Core::new().unwrap();
+            let client = Client::new(&core.handle());
+            for r in requests {
+                let work = client.request(r);
+                core.run(work).unwrap();
+            }
+        });
+    }
+
+    fn trigger(&self, gids: Vec<DataType>) {
```
Comment (jonhoo) [inline dataflow/src/ops/trigger.rs]: These are only g(roup)ids if the `TriggerType` is `GroupCreation`, right? Maybe just use `ids`?

File: dataflow/src/ops/union.rs [hunk 30]
```diff
@@ -9,9 +9,9 @@ enum Emit {
         emit: HashMap<IndexPair, Vec<usize>>,
 
         // generated
-        emit_l: Map<Vec<usize>>,
+        emit_l: BTreeMap<LocalNodeIndex, Vec<usize>>,
```
Comment (jonhoo) [inline dataflow/src/ops/union.rs]: This change was made because there are now a lot of IDs, right? Why a `BTreeMap` over a `HashMap`? I guess we expect there to be relatively few entries?

File: dataflow/src/processing.rs [hunk 31]
```diff
@@ -210,4 +210,9 @@ where
     fn is_selective(&self) -> bool {
         false
     }
+
+    /// Returns true if this operator should always materialize its state
+    fn must_materialize(&self) -> bool {
```
Comment (jonhoo) [inline dataflow/src/processing.rs]: As mentioned further up, I think this method should go away.

File: src/controller/migrate/materialization/mod.rs [hunk 32]
```diff
@@ -304,6 +304,10 @@ impl Materializations {
                 able = false;
             }
 
+            if graph[ni].is_internal() && graph[ni].must_materialize() {
```
Comment (jonhoo) [inline src/controller/migrate/materialization/mod.rs]: I'm confused by this. Is `must_materialize` used to indicate that *full* materialization is necessary?

File: src/controller/migrate/mod.rs [hunk 33]
```diff
@@ -117,6 +120,29 @@ impl<'a> Migration<'a> {
         ni.into()
     }
 
+    /// Returns the context of this migration
+    pub fn context(&self) -> HashMap<String, DataType> {
```
Comment (jonhoo) [inline src/controller/migrate/mod.rs]: Can this return `&HashMap` and not `clone` instead?

File: core/src/data.rs (hunk 8 above)
Comment (jonhoo) [reply core/src/data.rs]: Ah... I'm... I'd actually argue that the `impl Display` shouldn't quote strings. Quoted strings is in the domain of `impl Debug`. It feels unfortunate to have two basically identical ways of printing a `DataType`...

**Comment by jonhoo** [comment]
This is looking pretty good! I've done a pass over the more dataflow-y parts of the code, and left some comments. It'd probably be good to have @fintelia look over the controller changes, though nothing particularly egregious stood out to me there.

File: src/controller/migrate/materialization/mod.rs (hunk 32 above)
Comment (larat7) [reply src/controller/migrate/materialization/mod.rs]: Yes. The idea is that trigger nodes should always be fully materialized, otherwise group universes wouldn't exist until a user universe that belongs to that group is created. This might be something that we want to do, but it has some downsides: 1) it makes user universe creation slower, since it needs to wait for the necessary groups to be creates and 2) it's not super easy to guarantee order between migrations (the group universe needs to be created before the the user universe)

File: mir/src/optimize.rs (hunk 3 above)
Comment (larat7) [reply mir/src/optimize.rs]: Ah, I remember it was merging some filters that it shouldn't have and I meant to look into it, but ended up forgetting about it.

File: src/controller/inner.rs (hunk 5 above)
Comment (larat7) [reply src/controller/inner.rs]: Yeah, since the getter needs to know the security groups, I figured it was better to just handle it here instead of having `ControllerHandle` make multiple rpc calls (one to get the existing groups from the recipes and another to build the necessary getters).

File: dataflow/src/ops/rewrite.rs (hunk 23 above)
Comment (jonhoo) [reply dataflow/src/ops/rewrite.rs]: I believe this is still relevant?

File: src/controller/migrate/materialization/mod.rs (hunk 32 above)
Comment (jonhoo) [reply src/controller/migrate/materialization/mod.rs]: It also gets really awkward because you are now much more likely to end up with full-below-partial :/ If we do indeed need it though, the method should be renamed to `requires_full_materialization`, and there should be a comment in the impl for `Trigger` giving the rationale for why it has to be full.

File: dataflow/src/ops/trigger.rs (hunk 27 above)
Comment (larat7) [reply dataflow/src/ops/trigger.rs]: I imagine that if we want to trigger other events that are not just `GroupCreation` we might want a larger set of keys. Although for now they are all of length 1, since that's the only event we support.

File: src/controller/handle.rs [hunk 34]
```diff
@@ -185,6 +186,45 @@ impl ControllerHandle<LocalAuthority> {
...
+        self.rpc("enable_reuse", &reuse_type)
+    }
+
+    /// Set a fixed number of thread domains
+    pub fn set_ndomains(&mut self, ndomains: usize) {
+        self.rpc("set_ndomains", &ndomains)
+    }
+
+    /// Install a new set of policies on the controller.
+    pub fn set_security_config(&mut self, p: String) {
```
[the remark this answers is not in the document]
Comment (larat7) [reply src/controller/handle.rs]: Yeah, we have some scenarios where we would want to change the security configuration at runtime, similarly to the recipe.

File: dataflow/src/ops/trigger.rs (hunk 27 above)
Comment (jonhoo) [reply dataflow/src/ops/trigger.rs]: Then I think I'd just make it a `usize` for now, and we can worry about extending it further later. The extra heap allocation and `[0]`s aren't worth it.

File: dataflow/src/ops/union.rs (hunk 30 above)
Comment (larat7) [reply dataflow/src/ops/union.rs]: Yeah, this was changed because `Map` has the size of the biggest ID. I believe we discussed `HashMap` vs. `BTreeMap` and ended up deciding that the `BTreeMap` would be better, don't quite remember why though :(

File: dataflow/src/ops/rewrite.rs [hunk 35]
```diff
@@ -100,8 +100,12 @@ impl Ingredient for Rewrite {
 
 
                 if rc.is_none() {
-                    // todo (larat): might need to do some stuff here
-                    continue;
+                    misses.push(Miss {
+                       node: *self.signal,
+                       columns: vec![0],
+                       replay_key: None,
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: I *think* `replay_key` needs to be set here too.

Comment (larat7) [reply dataflow/src/ops/rewrite.rs]: Uhmm, so the `signal` parent is just used to check if the record should be rewritten or not, none of its columns actually make it into the output of the node (all columns come from the `src` parent).

Comment (jonhoo) [reply dataflow/src/ops/rewrite.rs]: But this is under `if from == *self.src`, right? Think of it this way: if we miss doing a lookup into `signal`, we'll want to first replay the key we missed on *to* `signal`, and then replay the key we were working on when we missed.

File: core/src/data.rs [hunk 36]
```diff
@@ -42,9 +42,9 @@ pub enum DataType {
     Timestamp(NaiveDateTime),
 }
 
-#[cfg(any(feature = "web", test))]
 impl DataType {
     /// Lossy representation as JSON value.
+    #[cfg(any(feature = "web", test))]
     pub fn to_json(&self) -> Value {
```
[the remark this answers is not in the document]
Comment (larat7) [reply core/src/data.rs]: I don't think this is being used anywhere else anymore. Should we just get rid of it? @ms705 @jonhoo

[the remark this answers is not in the document]
Comment (ms705) [reply core/src/data.rs]: I don't think we even have the `web` feature any more, so seems like it's good to go.

Only the `real_to_json` test in this file uses it, so let's drop that one too.

File: dataflow/src/ops/rewrite.rs (hunk 35 above)
Comment (larat7) [reply dataflow/src/ops/rewrite.rs]: I'm still a little confused about this. Would this mean that under `if from == *self.signal`, a miss in `src` would have a `None` `replay_key`? 

Otherwise, it seems weird that in both cases you can set `replay_key` to `replay_key_col.map(|col| vec![r[col].clone()])`, given what you said about `replay_key` being the column in the input that corresponds to the `replay_key_col` column in the output.

Comment (ms705) [reply dataflow/src/ops/rewrite.rs]: I think the situation is as follows: if we receive a record on `src`, and that record has a `key` value that we've never seen before (e.g., a Post ID), the partially-materialized `signal` parent will not have data for that key (the Post ID) and will need to query backwards up the graph to find out whether to rewrite the record or not. Hence, it needs to initiate a replay keyed on the Post ID that the corresponding column in the `src` input record holds.

Comment (larat7) [reply dataflow/src/ops/rewrite.rs]: Uhm, but then, how is `replay_key` different from `key` (which is also a parameter in `Miss)?

Comment (jonhoo) [reply dataflow/src/ops/rewrite.rs]: `key` is the key you used in the call to `lookup`. `replay_key` is the value of `replay_key_col` in the record you were processing when the lookup missed.

Comment (jonhoo) [reply dataflow/src/ops/rewrite.rs]: With the slight caveat that `replay_key_col` refers to an *output* column, so you have to trace it back to the input.

**Comment by larat7** [comment]
I think we might be ready to merge this. @jonhoo @ms705 @fintelia ?

File: dataflow/src/ops/rewrite.rs [hunk 37]
```diff
@@ -0,0 +1,242 @@
...
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+                if other_rows.is_none() {
+                     misses.push(Miss {
+                       node: *self.src,
+                       columns: vec![self.signal_key],
+                       replay_key: replay_key_col.map(|col| vec![r[col].clone()]),
```
Comment (jonhoo) [inline dataflow/src/ops/rewrite.rs]: In fact, I'm fairly sure that `replay_key_col` *must* be `None` in this case, since replays through these nodes always come from the `src` side, right?

Comment (ms705) [reply dataflow/src/ops/rewrite.rs]: Yes, they would always come from the source side -- a downstream miss requires data from the `src` side first for sure; however, it may subsequently also require data from the `signal` side to decide whether to rewrite (specifically, if it hasn't seen the key received from `src` before, I think).

Comment (jonhoo) [reply dataflow/src/ops/rewrite.rs]: I think we should check that this is indeed `None` here then, rather than do this (then incorrect) map.
